### PR TITLE
fix(frontend): `INSERT INTO t1 (v2, v1)` ignores specified columns list

### PR DESCRIPTION
**Pull request by CAJan93**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Ticket: https://github.com/risingwavelabs/risingwave/issues/4046

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [x] Test for incorrect column names 
- [x] Test for duplicate column names
- [x] `insert into t (v1) values (1, 5);` should give error 

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Ticket: https://github.com/risingwavelabs/risingwave/issues/4046

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by Gun9niR** [comment]
Insert target columns is ignored in `bind_insert`.

**Comment by Gun9niR** [comment]
Sorry for accidentally closing the PR. The target insert columns are ignored in `bind_insert`. IIUC, the backend does not support inserting into part of the columns in a user-specified order, because `InsertExecutor` always just takes the data chunk and writes it to the table source. 

I think we can add something like `column_ids` in `InsertExecutor`, so that we can reorder the columns in the data chunk and fill columns that are not being inserted into with empty columns.

File: src/batch/src/executor/insert.rs [hunk 1]
```diff
@@ -228,6 +248,7 @@ mod tests {
             source_manager.clone(),
             Box::new(mock_executor),
             "InsertExecutor".to_string(),
+            vec![], // Ignoring insertion order
```
Comment (CAJan93) [inline src/batch/src/executor/insert.rs:251]: I am ignoring the insertion order on purpose here, since we already validate that the insertion order works in the e2e tests. is that ok?

File: src/compute/tests/integration_tests.rs [hunk 2]
```diff
@@ -178,6 +178,7 @@ async fn test_table_materialize() -> StreamResult<()> {
         source_manager.clone(),
         insert_inner,
         "InsertExecutor".to_string(),
+        vec![], // ignore insertion order
```
Comment (CAJan93) [inline src/compute/tests/integration_tests.rs:181]: same here.

File: src/frontend/src/binder/insert.rs [hunk 3]
```diff
@@ -113,10 +116,55 @@ impl Binder {
             }
         };
 
+        let mut target_table_col_idxs: Vec<TableIdx> = vec![];
+        for query_column in &columns {
+            let column_name = &query_column.value; // value or real_value() ?
```
Comment (CAJan93) [inline src/frontend/src/binder/insert.rs:121]: I am not sure what to use: `value` or `real_value`. Could you help me out?

File: src/frontend/src/binder/insert.rs [hunk 4]
```diff
@@ -113,10 +116,55 @@ impl Binder {
...
+            let mut col_exists = false;
+            for (col_idx, table_column) in table_source.columns.iter().enumerate() {
+                // TODO: is there a better comparison then by col name?
+                if *column_name == table_column.name {
+                    target_table_col_idxs.push(TableIdx::from(col_idx));
+                    col_exists = true;
+                    break;
+                }
```
Comment (CAJan93) [inline src/frontend/src/binder/insert.rs:129]: Currently doing string comparison. Is there a better way?

File: src/common/src/catalog/mod.rs [hunk 5]
```diff
@@ -125,6 +125,57 @@ impl From<TableId> for u32 {
     }
 }
 
+#[derive(Clone, Copy, Debug, Display, Default, Hash, PartialOrd, PartialEq, Eq, Ord)]
+pub struct TableIdx {
+    pub table_id: u32,
```
Comment (BugenZhao) [inline src/common/src/catalog/mod.rs:130]: Why do we have an `id` inside the `Index` struct? 🤔

Comment (CAJan93) [reply src/common/src/catalog/mod.rs]: thank you for the catch. That was a copy-paste error.

File: src/common/src/catalog/mod.rs [hunk 6]
```diff
@@ -125,6 +125,57 @@ impl From<TableId> for u32 {
     }
 }
 
+#[derive(Clone, Copy, Debug, Display, Default, Hash, PartialOrd, PartialEq, Eq, Ord)]
+pub struct TableIdx {
+    pub table_idx: u32,
+}
```
Comment (chenzl25) [inline src/common/src/catalog/mod.rs:131]: It seems `ColumnIdx` is more suitable than `TableIdx`.

**Review by chenzl25 — CHANGES_REQUESTED** [review]

File: src/frontend/src/binder/insert.rs (hunk 3 above)
Comment (chenzl25) [reply src/frontend/src/binder/insert.rs:121]: Considering this case, we need to support case insensitive here. 
```sql 
create table t (a int, b int);
insert into t (A, B) values (1,2);
```

**Review by chenzl25 — CHANGES_REQUESTED** [review]

**Comment by fuyufjh** [comment]
Hi, any updates?

**Comment by CAJan93** [comment]
Thank you for pointing that out to me @chenzl25. I made the columns case-insensitive. 

@fuyufjh Was working on a cloud issue. Hope to get some work done here today

File: src/common/src/catalog/mod.rs (hunk 6 above)
Comment (CAJan93) [reply src/common/src/catalog/mod.rs]: You are absolutely right. I changed that.

File: src/frontend/src/binder/insert.rs (hunk 3 above)
Comment (CAJan93) [reply src/frontend/src/binder/insert.rs:121]: I think the easiest way to support case insensitivity is to just cast it to lower case, right? Other than that `real_value` is also related to case insensitivity. The docs of the function read: 

> Value after considering quote style In certain places, double quotes can force case-sensitive, but not always e.g. session variables.

Comment (CAJan93) [reply src/frontend/src/binder/insert.rs:121]: Using `real_value()` without casting to lower case also works for my test case. What do you prefer @chenzl25?

File: src/frontend/src/binder/insert.rs (hunk 4 above)
Comment (CAJan93) [reply src/frontend/src/binder/insert.rs]: I assume there is no better way. Marking this conversation as resolved

**Comment by chenzl25** [comment]
BTW, because you have changed the proto, you need to generate proto for dashboard manually.
https://github.com/risingwavelabs/risingwave/actions/runs/3427398602/jobs/5710350767
```
generated proto files are not up to date. Please run 'npm run gen-proto' at dashboard/ and commit the changes.
```

**Review by chenzl25 — APPROVED** [review]
LGTM

File: src/frontend/src/binder/insert.rs [hunk 7]
```diff
@@ -113,10 +116,54 @@ impl Binder {
...
 
+        let mut target_table_col_idxs: Vec<ColumnIdx> = vec![];
+        for query_column in &columns {
+            let column_name = &query_column.value; // value or real_value() ?
+            let mut col_exists = false;
+            for (col_idx, table_column) in table_source.columns.iter().enumerate() {
+                if *column_name.to_lowercase() == table_column.name.to_lowercase() {
+                    target_table_col_idxs.push(ColumnIdx::from(col_idx));
+                    col_exists = true;
+                    break;
```
Comment (BugenZhao) [inline src/frontend/src/binder/insert.rs:127]: How about directly `continue 'outer` here?

File: src/frontend/src/binder/insert.rs [hunk 8]
```diff
@@ -113,10 +116,54 @@ impl Binder {
             }
         };
 
+        let mut target_table_col_idxs: Vec<ColumnIdx> = vec![];
+        for query_column in &columns {
+            let column_name = &query_column.value; // value or real_value() ?
+            let mut col_exists = false;
+            for (col_idx, table_column) in table_source.columns.iter().enumerate() {
+                if *column_name.to_lowercase() == table_column.name.to_lowercase() {
```
Comment (BugenZhao) [inline src/frontend/src/binder/insert.rs:124]: Consider using `eq_ignore_ascii_case` here.

File: src/frontend/src/binder/insert.rs [hunk 9]
```diff
@@ -113,10 +116,54 @@ impl Binder {
...
+            std::cmp::Ordering::Less => (false, "INSERT has less target columns than values"),
+        };
+        if !eq_len && !target_table_col_idxs.is_empty() {
+            return Err(RwError::from(ErrorCode::BindError(msg.to_string())));
+        }
+
+        // Check if column was used multiple times in query e.g.
+        // insert into t (v1, v1) values (1, 5);
+        let mut uniq_cols = target_table_col_idxs.clone();
+        uniq_cols.dedup();
```
Comment (BugenZhao) [inline src/frontend/src/binder/insert.rs:155]: I'm afraid this method does not work as expected if the vec is not sorted. Do you want `all_unique`?

File: src/frontend/src/binder/insert.rs [hunk 10]
```diff
@@ -26,6 +27,8 @@ pub struct BoundInsert {
     /// Used for injecting deletion chunks to the source.
     pub table_source: BoundTableSource,
 
+    pub column_idxs: Vec<ColumnIdx>,
```
Comment (BugenZhao) [inline src/frontend/src/binder/insert.rs:30]: Could you please add some docs?

File: src/frontend/planner_test/tests/testdata/insert.yaml [hunk 11]
```diff
@@ -41,17 +74,18 @@
...
       └─BatchValues { rows: [['2020-01-01 01:02:03':Varchar::Timestamp::Time], ['03:04:05':Varchar::Time]] }
-- name: a `VALUES` without insert context may be invalid on its own (compare with
+- name:
+    a `VALUES` without insert context may be invalid on its own (compare with
     above)
   sql: |
     create table t (v1 time);
     values (timestamp '2020-01-01 01:02:03'), (time '03:04:05');
-  binder_error: 'Bind error: types Timestamp and Time cannot be matched'
+  binder_error: "Bind error: types Timestamp and Time cannot be matched"
```
Comment (BugenZhao) [inline src/frontend/planner_test/tests/testdata/insert.yaml:83]: Do you generate this file with `./risedev do-apply-planner-test`? 🤔

File: src/common/src/catalog/mod.rs [hunk 12]
```diff
@@ -126,6 +126,57 @@ impl From<TableId> for u32 {
     }
 }
 
+#[derive(Clone, Copy, Debug, Display, Default, Hash, PartialOrd, PartialEq, Eq, Ord)]
+pub struct ColumnIdx {
```
Comment (BugenZhao) [inline src/common/src/catalog/mod.rs:130]: This struct seems irrelevant to the module `catalog`. 😢 Can we simply use the bare `usize` for the column index, which is widely used in other occurrences?

File: src/common/src/catalog/mod.rs [hunk 13]
```diff
@@ -126,6 +126,57 @@ impl From<TableId> for u32 {
...
+    pub column_idx: u32,
```
Comment (BugenZhao) [inline src/common/src/catalog/mod.rs:131]: We generally use `usize` for the indices in Rust.

File: src/frontend/src/binder/insert.rs (hunk 9 above)
Comment (BugenZhao) [reply src/frontend/src/binder/insert.rs:155]: For example, could you please test whether `insert into t (v1, v2, v1)` works?

File: src/frontend/src/binder/values.rs [hunk 14]
```diff
@@ -92,9 +92,17 @@ impl Binder {
         // Calculate column types.
         let types = match expected_types {
             Some(types) => {
+                if types.len() > num_columns {
```
Comment (BugenZhao) [inline src/frontend/src/binder/values.rs:95]: Is this check necessary? I guess `cast_on_insert` will do this for us.

File: src/frontend/src/binder/insert.rs (hunk 7 above)
Comment (CAJan93) [reply src/frontend/src/binder/insert.rs:127]: That is a nice idea. let me do that.

File: src/frontend/src/binder/insert.rs (hunk 8 above)
Comment (CAJan93) [reply src/frontend/src/binder/insert.rs:124]: Good advice 👍

File: src/frontend/src/binder/insert.rs (hunk 9 above)
Comment (CAJan93) [reply src/frontend/src/binder/insert.rs:155]: So I believe for the above example this should still work, because you then define more columns than there are in the table: 

```
insert into t1 (v1, v1) values (1, 5);
ERROR:  QueryError: Bind error: Column specified more than once
dev=>  insert into t1 (v1, v2, v1) values (1, 5);
ERROR:  QueryError: Bind error: INSERT has more target columns than values
```

However of cause you are right: It does not work as extended. Will change it. Thank you for the catch.

File: src/frontend/planner_test/tests/testdata/insert.yaml (hunk 11 above)
Comment (CAJan93) [reply src/frontend/planner_test/tests/testdata/insert.yaml:83]: No. I wrote this manually and on a different occasion I wrote a python scripts that writes tests for me. Why? Is there an issue with the quoting?

Comment (CAJan93) [reply src/frontend/planner_test/tests/testdata/insert.yaml:83]: Just tried the tool. Will upload the tool-version to stay consistent with the "code style" of the tests.

Comment (BugenZhao) [reply src/frontend/planner_test/tests/testdata/insert.yaml:83]: Not a big deal. I guess it will be overridden with some next PRs. 🥰

File: src/common/src/catalog/mod.rs (hunk 12 above)
Comment (CAJan93) [reply src/common/src/catalog/mod.rs:130]: Of cause we can :D I only created `ColumnIdx` to give future coders an easy understanding of what this represents. `usize` works just as well for me.

File: src/frontend/planner_test/tests/testdata/insert.yaml (hunk 11 above)
Comment (CAJan93) [reply src/frontend/planner_test/tests/testdata/insert.yaml]: Still. Good catch 👍

File: src/common/src/catalog/mod.rs (hunk 13 above)
Comment (CAJan93) [reply src/common/src/catalog/mod.rs]: I believe after removing `ColumnIdx` I am using `usize` for the indexes 🤔

File: src/frontend/src/binder/values.rs (hunk 14 above)
Comment (CAJan93) [reply src/frontend/src/binder/values.rs]: Yep, you are correct. I am a bit surprised tbh. During development I was running into issues without this check, but I ran it and it seems to be fine.

**Comment by CAJan93** [comment]
I hope I caught all the comments. Please let me know what you think @BugenZhao and @chenzl25. Thank you for all your support :)

**Comment by chenzl25** [comment]
> I hope I caught all the comments. Please let me know what you think @BugenZhao and @chenzl25. Thank you for all your support :)

Please make sure all the CI test cases have passed. Currently some unit tests and `dashboard-ui-build` still fail.

**Comment by fuyufjh** [comment]
> > I hope I caught all the comments. Please let me know what you think @BugenZhao and @chenzl25. Thank you for all your support :)
> 
> Please make sure all the CI test cases have passed. Currently some unit tests and `dashboard-ui-build` still fail.

CI failed due to multiple reasons

```
  | Test #9 (id: <none>) failed, SQL:
  | create table t (v1 real, v2 int, v3 varchar);
  | insert into t values (1, 2), (3, 4);
  | Error: Expected binder error: Bind error: VALUES list must define a value for each column in table
  | Actual binder error: Bind error: INSERT has more target columns than expressions
```

```
error: useless conversion to the same type: `usize`
--
  | --> src/batch/src/executor/insert.rs:101:47
  | \|
  | 101 \|                     ordered_cols.push(columns[usize::from(*idx)].clone());
  | \|                                               ^^^^^^^^^^^^^^^^^ help: consider removing `usize::from()`: `(*idx)`
  | \|
  | = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
  | = note: `-D clippy::useless-conversion` implied by `-D warnings`

```

For the dashboard stuff, please install `npm` locally and follow the instructions `npm run gen-proto`

**Review by BugenZhao — APPROVED** [review]
LGTM. Good work! Please fix the clippy issue with `./risedev check`.
